subprojects/build-init/src/main/groovy/org/gradle/api/tasks/wrapper/Wrapper.java

```diff
@@ -63,6 +64,7 @@
     private PathBase distributionBase = PathBase.GRADLE_USER_HOME;
     private String distributionUrl;
     private GradleVersion gradleVersion;
+    private String distType;
```

Comment:
It is not perfect, but good enough for now. It is still a little bit nicer of you change the type of the setter and the getter to `DistributionType`. The command line parser still can parse this. You can then remove the error handling from the setter. The error message if something not acceptable is provided will be:
```
Problem configuring option 'dist-type' on task ':wrapper' from command line.
> Cannot convert string value 'invalid-distribution-type' to an enum value of type 'org.gradle.api.tasks.wrapper.Wrapper$DistributionType' (valid case insensitive values: BIN, ALL)
```
which is good enough.

Moreover, please add a method
```groovy
@OptionValues("dist-type")
public List<DistributionType> availableDistTypes() {
    return Arrays.asList(DistributionType.values());
}
```

This will cause that running `gradle help --task wrapper` yields:
```
Type
     Wrapper (org.gradle.api.tasks.wrapper.Wrapper)

Options
     --dist-type     The type of the Gradle distribution to be used by the wrapper.
                     Available values are:
                          ALL
                          BIN

     --gradle-distribution-url     The URL to download the Gradle distribution from.

     --gradle-version     The version of the Gradle distribution required by the wrapper.
```